nix/pwndbg.nix

```
@@ -35,33 +40,61 @@ let
     ''
   );
 
-  pwndbg = pkgs.stdenv.mkDerivation {
-    name = "pwndbg";
+  pwndbg = let
+    pwndbgName = if isLLDB then "pwndbg-lldb" else "pwndbg";
+  in pkgs.stdenv.mkDerivation {
+    name = pwndbgName;
     version = pwndbgVersion;
 
-    src = pkgs.lib.sourceByRegex inputs.pwndbg [
+    src = pkgs.lib.sourceByRegex inputs.pwndbg ([
       "pwndbg"
       "pwndbg/.*"
+    ] ++ (if isLLDB then [
+      "lldbinit.py"
+      "pwndbg-lldb.py"
+    ] else [
       "gdbinit.py"
-    ];
+    ]));
 
     nativeBuildInputs = [ pkgs.makeWrapper ];
 
-    installPhase = ''
+    installPhase = let
+      fix_init_script = { target, line }: ''
+        # Build self-contained init script for lazy loading from vanilla gdb
+        # I purposely use insert() so I can re-import during development without having to restart gdb
+        sed "${line} i import sys, os\n\
+        sys.path.insert(0, '${pyEnv}/${pyEnv.sitePackages}')\n\
+        sys.path.insert(0, '$out/share/pwndbg/')\n\
+        os.environ['PATH'] += ':${binPath}'\n" -i ${target}
+      '';
+    in (if isLLDB then ''
+      mkdir -p $out/share/pwndbg
+      mkdir -p $out/bin
+
+      cp -r lldbinit.py pwndbg $out/share/pwndbg
+      cp pwndbg-lldb.py $out/bin/${pwndbgName}
+
+      # patchShebangs isn't working, so we do it manually.
+      sed -i "1 d" $out/bin/${pwndbgName}
+      sed -i "1 i #!${pkgs.lib.makeBinPath [ python3 ]}/python3" $out/bin/${pwndbgName}
+
+      ${fix_init_script { target = "$out/bin/${pwndbgName}"; line = "4"; } }
+
+      touch $out/share/pwndbg/.skip-venv
+      wrapProgram $out/bin/${pwndbgName} \
+        --prefix PATH : ${ pkgs.lib.makeBinPath [ lldb ] } \
+        --set PWNDBG_LLDBINIT_DIR $out/share/pwndbg \
+        --set LLDB_DEBUGSERVER_PATH ${ pkgs.lib.makeBinPath [ lldb ] }/lldb-server
```

```suggestion
      wrapProgram $out/bin/${pwndbgName} \
        --prefix PATH : ${ pkgs.lib.makeBinPath [ lldb ] } \
        '' + (pkgs.lib.optionalString !pkgs.stdenv.isDarwin ''
        --set-default LLDB_DEBUGSERVER_PATH ${ pkgs.lib.makeBinPath [ lldb ] }/lldb-server \
        '') + ''
        --set PWNDBG_LLDBINIT_DIR $out/share/pwndbg
```
- for darwin `LLDB_DEBUGSERVER_PATH` env must be unset
- for linux would be nice to use `set-default` it allows overriding 

it looks little ugly

---

Yeah, it does look ugly for sure. Also, we probably don't want to let Linux users override it, since, in general, we probably want the debug server to always match the LLDB package we use. Darwin is an exception, because we have no other choice. Overriding the variable should probably be done by customizing the Nix files or overriding the derivation.

---

Ok 👍